File: clients/android/app/src/main/java/com/onlineclipboard/app/core/Crypto.kt
package com.onlineclipboard.app.core

import android.util.Base64
import java.nio.charset.StandardCharsets
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import org.bouncycastle.crypto.generators.Argon2BytesGenerator
import org.bouncycastle.crypto.params.Argon2Parameters

object CryptoV1 {
    fun encode(raw: ByteArray): String =
        Base64.encodeToString(raw, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

    fun decode(value: String): ByteArray {
        val raw = Base64.decode(value, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
        if (encode(raw) != value) throw IllegalArgumentException("non-canonical base64url")
        return raw
    }

    fun random(n: Int): ByteArray {
        val buf = ByteArray(n)
        java.security.SecureRandom().nextBytes(buf)
        return buf
    }

    fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int = 32): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        val saltKey = if (salt.isEmpty()) ByteArray(32) else salt
        mac.init(SecretKeySpec(saltKey, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        mac.update(info)
        mac.update(0x01)
        return mac.doFinal().copyOf(length)
    }

    fun aesGcm(key: ByteArray, nonce: ByteArray, plaintext: ByteArray, aad: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        c.updateAAD(aad)
        return c.doFinal(plaintext)
    }

    fun aesGcmOpen(key: ByteArray, nonce: ByteArray, ciphertext: ByteArray, aad: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        c.updateAAD(aad)
        return c.doFinal(ciphertext)
    }

    fun recoveryCode(rk: ByteArray) = "oc1_" + encode(rk)

    fun parseRecovery(code: String): ByteArray {
        val compact = code.replace(" ", "")
        require(compact.startsWith("oc1_")) { "recovery code must start with oc1_" }
        val raw = decode(compact.removePrefix("oc1_"))
        require(raw.size == 32) { "recovery key must be 32 bytes" }
        return raw
    }

    const val passwordKdfTime = 3
    const val passwordKdfMemoryKib = 64 * 1024
    const val passwordKdfParallel = 1

    fun wrapCmk(rk: ByteArray, salt: ByteArray, nonce: ByteArray, cmk: ByteArray, userId: String, vaultId: String): ByteArray {
        val kek = hkdf(rk, salt, "onlineclipboard/v1/wrap".toByteArray(StandardCharsets.UTF_8))
        val aad = "oc-v1|vault|$userId|$vaultId|1".toByteArray(StandardCharsets.UTF_8)
        return aesGcm(kek, nonce, cmk, aad)
    }

    fun unwrapCmk(rk: ByteArray, salt: ByteArray, nonce: ByteArray, wrapped: ByteArray, userId: String, vaultId: String): ByteArray {
        val kek = hkdf(rk, salt, "onlineclipboard/v1/wrap".toByteArray(StandardCharsets.UTF_8))
        val aad = "oc-v1|vault|$userId|$vaultId|1".toByteArray(StandardCharsets.UTF_8)
        return aesGcmOpen(kek, nonce, wrapped, aad)
    }

    fun passwordIkm(password: String, salt: ByteArray, time: Int, memory: Int, parallel: Int): ByteArray {
        val t = if (time <= 0) passwordKdfTime else time
        val m = if (memory <= 0) passwordKdfMemoryKib else memory
        val p = if (parallel <= 0) passwordKdfParallel else parallel
        val params = Argon2Parameters.Builder(Argon2Parameters.ARGON2_id)
            .withVersion(Argon2Parameters.ARGON2_VERSION_13)
            .withIterations(t)
            .withMemoryAsKB(m)
            .withParallelism(p)
            .withSalt(salt)
            .build()
        val gen = Argon2BytesGenerator()
        gen.init(params)
        val out = ByteArray(32)
        gen.generateBytes(password.toByteArray(StandardCharsets.UTF_8), out)
        return out
    }

    fun wrapCmkWithPassword(
        password: String, kdfSalt: ByteArray, wrapSalt: ByteArray, wrapNonce: ByteArray,
        cmk: ByteArray, userId: String, vaultId: String
    ): ByteArray {
        val kek = hkdf(
            passwordIkm(password, kdfSalt, passwordKdfTime, passwordKdfMemoryKib, passwordKdfParallel),
            wrapSalt,
            "onlineclipboard/v1/wrap-password".toByteArray(StandardCharsets.UTF_8)
        )
        val aad = "oc-v1|vault-password|$userId|$vaultId|1".toByteArray(StandardCharsets.UTF_8)
        return aesGcm(kek, wrapNonce, cmk, aad)
    }

    fun unwrapCmkWithPassword(
        password: String, kdfSalt: ByteArray, wrapSalt: ByteArray, wrapNonce: ByteArray,
        wrapped: ByteArray, userId: String, vaultId: String, time: Int, memory: Int, parallel: Int
    ): ByteArray {
        val kek = hkdf(
            passwordIkm(password, kdfSalt, time, memory, parallel),
            wrapSalt,
            "onlineclipboard/v1/wrap-password".toByteArray(StandardCharsets.UTF_8)
        )
        val aad = "oc-v1|vault-password|$userId|$vaultId|1".toByteArray(StandardCharsets.UTF_8)
        return aesGcmOpen(kek, wrapNonce, wrapped, aad)
    }

    fun makePasswordWrap(password: String, cmk: ByteArray, userId: String, vaultId: String): PasswordWrap {
        val kdfSalt = random(16)
        val wrapSalt = random(32)
        val nonce = random(12)
        val wrapped = wrapCmkWithPassword(password, kdfSalt, wrapSalt, nonce, cmk, userId, vaultId)
        return PasswordWrap(
            time = passwordKdfTime,
            memory = passwordKdfMemoryKib,
            parallelism = passwordKdfParallel,
            kdfSalt = encode(kdfSalt),
            wrapSalt = encode(wrapSalt),
            wrapNonce = encode(nonce),
            wrappedKey = encode(wrapped)
        )
    }

    fun encryptItem(
        cmk: ByteArray, vaultId: String, userId: String, clipId: String, sourceDeviceId: String,
        epoch: Int, nonce: ByteArray, plaintext: ByteArray
    ): ByteArray {
        val info = "onlineclipboard/v1/item|$clipId|$epoch".toByteArray(StandardCharsets.UTF_8)
        val key = hkdf(cmk, vaultId.toByteArray(StandardCharsets.UTF_8), info)
        val aad = "oc-v1|clip|$userId|$vaultId|$clipId|$sourceDeviceId|$epoch|text/plain".toByteArray(StandardCharsets.UTF_8)
        return aesGcm(key, nonce, plaintext, aad)
    }

    fun decryptItem(
        cmk: ByteArray, vaultId: String, userId: String, clipId: String, sourceDeviceId: String,
        epoch: Int, nonce: ByteArray, ciphertext: ByteArray
    ): ByteArray {
        val info = "onlineclipboard/v1/item|$clipId|$epoch".toByteArray(StandardCharsets.UTF_8)
        val key = hkdf(cmk, vaultId.toByteArray(StandardCharsets.UTF_8), info)
        val aad = "oc-v1|clip|$userId|$vaultId|$clipId|$sourceDeviceId|$epoch|text/plain".toByteArray(StandardCharsets.UTF_8)
        return aesGcmOpen(key, nonce, ciphertext, aad)
    }

    fun uuid(): String = java.util.UUID.randomUUID().toString()
}

data class PasswordWrap(
    val kdf: String = "argon2id",
    val time: Int,
    val memory: Int,
    val parallelism: Int,
    val kdfSalt: String,
    val wrapSalt: String,
    val wrapNonce: String,
    val wrappedKey: String
)

File: clients/windows/OnlineClipboard.Windows/Core/Crypto.cs
using System.Security.Cryptography;
using System.Text;

namespace OnlineClipboard.Windows.Core;

public static class CryptoV1
{
    public const int FormatVersion = 1;
    public const int KeyEpoch = 1;
    private static readonly byte[] WrapInfo = Encoding.UTF8.GetBytes("onlineclipboard/v1/wrap");

    public static byte[] RandomBytes(int n)
    {
        var buf = new byte[n];
        RandomNumberGenerator.Fill(buf);
        return buf;
    }

    public static string Encode(byte[] raw) => Convert.ToBase64String(raw).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    public static byte[] Decode(string value)
    {
        var padded = value.Replace('-', '+').Replace('_', '/');
        switch (padded.Length % 4)
        {
            case 2: padded += "=="; break;
            case 3: padded += "="; break;
            case 1: throw new FormatException("non-canonical base64url");
        }
        var raw = Convert.FromBase64String(padded);
        if (Encode(raw) != value) throw new FormatException("non-canonical base64url");
        return raw;
    }

    public static string RecoveryCode(byte[] rk) => "oc1_" + Encode(rk);

    public static byte[] ParseRecoveryCode(string code)
    {
        code = code.Replace(" ", "", StringComparison.Ordinal);
        if (!code.StartsWith("oc1_", StringComparison.Ordinal)) throw new FormatException("recovery code must start with oc1_");
        var raw = Decode(code[4..]);
        if (raw.Length != 32) throw new FormatException("recovery key must be 32 bytes");
        return raw;
    }

    public static byte[] Derive(byte[] ikm, byte[] salt, byte[] info) =>
        HKDF.DeriveKey(HashAlgorithmName.SHA256, ikm, 32, salt, info);

    public static byte[] AesGcmSeal(byte[] key, byte[] nonce, byte[] plaintext, byte[] aad)
    {
        var cipher = new byte[plaintext.Length];
        var tag = new byte[16];
        using var gcm = new AesGcm(key, 16);
        gcm.Encrypt(nonce, plaintext, cipher, tag, aad);
        var combined = new byte[cipher.Length + tag.Length];
        Buffer.BlockCopy(cipher, 0, combined, 0, cipher.Length);
        Buffer.BlockCopy(tag, 0, combined, cipher.Length, tag.Length);
        return combined;
    }

    public static byte[] AesGcmOpen(byte[] key, byte[] nonce, byte[] ciphertextAndTag, byte[] aad)
    {
        if (ciphertextAndTag.Length < 17) throw new CryptographicException("ciphertext too short");
        var cipherLen = ciphertextAndTag.Length - 16;
        var cipher = ciphertextAndTag[..cipherLen];
        var tag = ciphertextAndTag[cipherLen..];
        var plain = new byte[cipherLen];
        using var gcm = new AesGcm(key, 16);
        gcm.Decrypt(nonce, cipher, tag, plain, aad);
        return plain;
    }

    public static byte[] WrapAad(string userId, string vaultId) =>
        Encoding.UTF8.GetBytes($"oc-v1|vault|{userId}|{vaultId}|1");

    public static byte[] ItemInfo(string clipId, int epoch) =>
        Encoding.UTF8.GetBytes($"onlineclipboard/v1/item|{clipId}|{epoch}");

    public static byte[] ItemAad(string userId, string vaultId, string clipId, string sourceDeviceId, int epoch) =>
        Encoding.UTF8.GetBytes($"oc-v1|clip|{userId}|{vaultId}|{clipId}|{sourceDeviceId}|{epoch}|text/plain");

    public static byte[] WrapCmk(byte[] rk, byte[] wrapSalt, byte[] wrapNonce, byte[] cmk, string userId, string vaultId)
    {
        var kek = Derive(rk, wrapSalt, WrapInfo);
        return AesGcmSeal(kek, wrapNonce, cmk, WrapAad(userId, vaultId));
    }

    public static byte[] UnwrapCmk(byte[] rk, byte[] wrapSalt, byte[] wrapNonce, byte[] wrapped, string userId, string vaultId)
    {
        var kek = Derive(rk, wrapSalt, WrapInfo);
        return AesGcmOpen(kek, wrapNonce, wrapped, WrapAad(userId, vaultId));
    }

    public static byte[] EncryptItem(byte[] cmk, string vaultId, string userId, string clipId, string sourceDeviceId, int epoch, byte[] nonce, byte[] plaintext)
    {
        var key = Derive(cmk, Encoding.UTF8.GetBytes(vaultId), ItemInfo(clipId, epoch));
        return AesGcmSeal(key, nonce, plaintext, ItemAad(userId, vaultId, clipId, sourceDeviceId, epoch));
    }

    public static byte[] DecryptItem(byte[] cmk, string vaultId, string userId, string clipId, string sourceDeviceId, int epoch, byte[] nonce, byte[] ciphertext)
    {
        var key = Derive(cmk, Encoding.UTF8.GetBytes(vaultId), ItemInfo(clipId, epoch));
        return AesGcmOpen(key, nonce, ciphertext, ItemAad(userId, vaultId, clipId, sourceDeviceId, epoch));
    }

    public static byte[] ProtectLocal(byte[] data) => ProtectedData.Protect(data, null, DataProtectionScope.CurrentUser);
    public static byte[] UnprotectLocal(byte[] data) => ProtectedData.Unprotect(data, null, DataProtectionScope.CurrentUser);
}
